cmd/clusterctl/pkg/client/cluster/upgrader.go

```diff
@@ -99,40 +109,250 @@ func (u *providerUpgrader) Plan() ([]UpgradePlan, error) {
 		// an UpgradeItem for each provider defining the next available version with the target contract, if available.
 		// e.g. v1alpha3, cluster-api --> v0.3.2, kubeadm bootstrap --> v0.3.2, aws --> v0.5.4
 		// e.g. v1alpha4, cluster-api --> v0.4.1, kubeadm bootstrap --> v0.4.1, aws --> v0.6.2
-		for _, contract := range contractsForUpgrade {
-			upgradeItems := []UpgradeItem{}
-			for _, provider := range managementGroup.Providers {
-				// Gets the upgrade info for the provider.
-				providerUpgradeInfo, err := u.getUpgradeInfo(provider)
-				if err != nil {
-					return nil, err
-				}
-
-				// Identifies the next available version with the target contract for the provider, if available.
-				nextVersion := providerUpgradeInfo.getLatestNextVersion(contract)
-
-				// Append the upgrade item for the provider/with the target contract.
-				upgradeItems = append(upgradeItems, UpgradeItem{
-					Provider:    provider,
-					NextVersion: versionTag(nextVersion),
-				})
+		for _, apiVersion := range contractsForUpgrade {
+			upgradePlan, err := u.getUpgradePlan(managementGroup, apiVersion)
+			if err != nil {
+				return nil, err
 			}
-
-			ret = append(ret, UpgradePlan{
-				Contract:     contract,
-				CoreProvider: managementGroup.CoreProvider,
-				Providers:    upgradeItems,
-			})
+			ret = append(ret, *upgradePlan)
 		}
 	}
 
 	return ret, nil
 }
 
-func newProviderUpgrader(configClient config.Client, repositoryClientFactory RepositoryClientFactory, providerInventory InventoryClient) *providerUpgrader {
+func (u *providerUpgrader) ApplyPlan(coreProvider clusterctlv1.Provider, contract string) error {
+	log := logf.Log
+	log.Info("Performing upgrade...")
```

Comment:
Can we pass the logger in from the caller and log this one out in `ApplyUpgrade` instead?

Reply:
Uhmm, the whole idea around the implementation of a global logger was to avoid to pass the logger object through the call chain... If possible I prefer to keep this consistent with the rest of the code base